pkg/secure/mqttfactory.go

```diff
@@ -110,3 +106,47 @@ func (factory MqttFactory) configureMQTTClientForAuth(secretData *messaging.Secr
 
 	return nil
 }
+
+func (factory MqttFactory) validateSecretData() (*messaging.SecretData, error) {
+	type result struct {
+		secretData *messaging.SecretData
+		err        error
+	}
+
+	f := func(channel chan result) {
+		//get the secrets from the secret provider and populate the struct
+		secretData, err := messaging.GetSecretData(factory.authMode, factory.secretPath, factory.sp)
+		if err != nil {
+			channel <- result{nil, err}
+			return
+		}
+		//ensure that the authmode selected has the required secret values
+		err = messaging.ValidateSecretData(factory.authMode, factory.secretPath, secretData)
+		if err != nil {
+			channel <- result{nil, err}
+		} else {
+			channel <- result{secretData, nil}
+		}
+	}
+
+	var ch = make(chan result)
+	defer close(ch)
+
+	go f(ch)
+	for {
+		select {
+		case r := <-ch:
+			if r.err != nil {
+				factory.logger.Errorf("failed to get valid secret data, error: %s. ", r.err)
+				if factory.secretAddedSignal == nil || !secret.IsSecurityEnabled() {
+					return nil, r.err
+				}
+				factory.logger.Debug("Waiting for a SecretAdded signal...")
+			} else {
+				return r.secretData, nil
+			}
+		case <-factory.secretAddedSignal:
+			go f(ch)
```

Comment:
Note that if more than one of the channels are ready, go **randomly** picks which one to receive from.  Under such case, is it possible to meet race condition for this select block?

Reply:
Good catch, thanks. No need to use channel here.